### ISSUE TITLE
Extract is broken

### ISSUE THREAD

**Issue by titpetric** [h2non/imaginary#127]
In documentation:

```
width int - Width of image area to extract/resize
height int - Height of image area to extract/resize
top int - Top edge of area to extract. Example: 100
left int - Left edge of area to extract. Example: 100
areawidth int - Height area to extract. Example: 300
areaheight int - Width area to extract. Example: 300
```

So, if we have for example this image:

![freud-cigar](https://cloud.githubusercontent.com/assets/233360/23709256/206c71ba-0419-11e7-8623-af4ea95aca69.jpg)

I would asume that top/left/areawidth/areaheight are applied before any width/height transformation.
So, very valid parameters for this 350 x 462 image would be:

`top=32&left=0&areawidth=350&areaheight=350&width=200&height=200`

These parameters on the extract endpoint should produce a 200x200 image, which is a resized extracted area (scaled down from 350 x 350). In reality however:

`{"message":"Error while processing the image: extract_area: bad extract area","code":1}`

So, it seems if areawidth/height is larger than height/width the above error occurs. The same error occurs if I set height/width to 350x350 - because of top=32. So I set these to the natural dimensions of 350x462.

![image](https://cloud.githubusercontent.com/assets/233360/23709572/2fcd9a5c-041a-11e7-8dca-3edc82b7f031.png)

Works as intended. So, it might be that I need to decrease areawidth/height to get a resized image? Seems like no:

![image](https://cloud.githubusercontent.com/assets/233360/23709621/562e6208-041a-11e7-9c68-da6623de4e63.png)

The thing is, if the image is resized, the crop area should be proportionately resized based on the original image. The operation of extract seems to be resize first, extract after - while the more logical way would be to extract first, and resize after.

**Comment by scott-seo** [comment]
I was looking at imaginary, bimg, and libvips code. It seems to me this might be a matter of either updating the documentation on imaginary. the code that deals with extracting only expects four parameters, left, top, areawidth and areaheight. if width and height are passed in resizing takes place first which causes areawidth to be more than the newly resized width. that's why "bad extract area" error occurs.

I am able to get around this problem by only using the expected four parameters on the first pass,  then resize on the second pass.

This is the unencoded version for reading.

http://localhost:9000/resize?width=XX&height=YY&url=http://localhost:9000/extract?left=XX&top=XX&areawidth=XX&areaheight&url=http://foo/img.jpg

Actual encoded URL will be this
http://localhost:9000/resize?width=XX&height=YY&url=http%3A%2F%2Flocalhost%3A9000%2Fextract%3Fleft%3DXX%26top%3DXX%26areawidth%3DXX%26areaheight%26url%3Dhttp%3A%2F%2Ffoo%2Fimg.jpg

This uses a bit of IO but it's not terrible.

I am not using imaginary directly. I am using bimg directly in my code. I do extract first then resize.

@h2non I think I can make this change and do a pull request if you think my approach is the right way to go.

**Comment by h2non** [comment]
Ideally speaking `imaginary` should be smart enough and infer that behavior internally by transparently doing both `extract` and `resize` operations accordingly.

Two HTTP requests is not great but not a big deal IMO. @scott-seo I'm fully open to merge any PR that can help other users understanding the current behavior, so other users can come up with a similar approach. However, implementing resize-extract operations multiplexing should be supported once #101 is implemented.

**Comment by titpetric** [comment]
I'll flag this into won't fix territory and mark it as closed as #101 was implemented.
